tests/virt_autotest/vgpu.pm

```diff
@@ -0,0 +1,368 @@
+# SUSE's openQA tests
+# Copyright 2020 SUSE LLC
+# SPDX-License-Identifier: GPL-2.0-or-later
+#
+# Summary: A test to pass an NVIDIA vGPU to guest via libvirt. Only KVM hosts are supported.
+# Test environment: one AMPERE GPU card, such A10, A30 etc., with SR-IOV capability and/or MIG features in host machine;
+#                   A UEFI guest vm is defined in the host, and they are ssh accessible from host.
+# Test flow:
+#    - install the NVIDIA vGPU manager on host and reboot
+#    - create a GPU Instance and a Compute Instance if MIG feature is available
+#    - create a vGPU device in host
+#    - assign the vGPU device to guest vm
+#    - install the vGPU driver in guest vm
+#    - detach vGPU from guest and remove the vGPU
+# Maintainer: qe-virt@suse.de, Julie CAO <JCao@suse.com>
+
+use base "virt_feature_test_base";
+use strict;
+use warnings;
+use utils;
+use testapi;
+use virt_autotest::common;
+use version_utils qw(is_sle);
+use virt_autotest::utils;
+use Utils::Backends qw(use_ssh_serial_console is_remote_backend);
+use ipmi_backend_utils qw(reconnect_when_ssh_console_broken);
+
+our $log_dir = "/tmp/vgpu";
+our $vm_xml_save_dir = "/tmp/download_vm_xml";
+
+sub run_test {
+    my $self = shift;
+
+    # Use sol & ssh console explixitly in case test module run on an installed host
+    # thus no login_console.pm has been required any more
+    select_console 'sol', await_console => 0;
+    use_ssh_serial_console;
+
+    # Test Prerequisite is fulfilled
+    return unless is_host_ready_for_vgpu_test() eq 'yes';
+
+    my $gpu_device = get_gpu_device();
+    die "No NVIDIA AMPERE GPU card on the machine!" if $gpu_device eq '';
+
+    # Clean up test logs
+    script_run "[ -d $log_dir ] && rm -rf $log_dir; mkdir -p $log_dir";
+
+    # Install NVIDIA vGPU manager
+    my $vgpu_manager = get_required_var("VGPU_MANAGER_URL");
+    install_vgpu_manager_and_reboot($vgpu_manager);
+
+    # Enable SR-IOV VFs
+    enable_sriov($gpu_device);
+
+    return if get_var('VGPU_SHORT_TEST');
+
+    # Create a CI for GPU with MIG mode supported, such A30/A100
+    # given MIG mode is enabled. will handle non MIG mode when zen2 is available
+    if (get_var('GPU_MIG_MODE')) {
```

Comment:
Is it possible to detect which mode is supported via checking gpu info in test code, for MIG or non-MIG?

Reply:
After installing NVIDIA vgpu driver, it is not difficult to distinguish if MIG mode is supported for this type of GPU device. but whether to switch MIG mode on or off, it depends on our test purpose, so I leave it switchable.

Reply:
I see. Sounds good.